src/core.jl

```diff
@@ -175,7 +175,16 @@ output and `ȳ` the reverse-mode sensitivity of `y`.
 ∇(y::Node, ȳ) = propagate(tape(y), reverse_tape(y, ȳ))
 @inline ∇(y::Node{<:∇Scalar}) = ∇(y, one(unbox(y)))
 
-@inline ∇(x̄, f, ::Type{Arg{N}}, args...) where N = x̄ + ∇(f, Arg{N}, args...)
+# This is a fallback method where we don't necessarily know what we'll be adding and whether
+# we can update the value in-place, so we'll try to be clever and dispatch.
+@inline ∇(x̄, f, ::Type{Arg{N}}, args...) where {N} = update!(x̄, ∇(f, Arg{N}, args...))
+
+# Update arrays in-place. Mixed array and scalar adds should not occur, so don't bother
+# accepting scalars on the RHS.
+@inline update!(x̄::∇Array, y::∇Array) = x̄ .+= y
```

Comment:
I realized that this is not legit to do in the general case; we'll have to be even more restrictive. Say we have an `UpperTriangular` stored on the tape. When we go to update this, if we do `.+=`, the resulting value is still an `UpperTriangular`, which is wrong. I think updating the value on the tape in-place is _only_ safe when the existing and new values are both `Array`s.